modules/imap/functions.php

```diff
@@ -1425,3 +1425,50 @@ function parse_sieve_config_host($host) {
     // $host = '$scheme://'.$host;
     return [$host, $port, $tls];
 }}
+
+if (!hm_exists('connect_to_imap_server')) {
+    function connect_to_imap_server($address, $name, $port, $user, $pass, $tls, $imap_sieve_host, $enableSieve, $type, $context, $hidden = false) {
+        Hm_IMAP_List::init($context->user_config, $context->session);
+        $imap_list = array(
+            'name' => $name,
+            'server' => $address,
+            'hide' => false,
+            'port' => $port,
+            'user' => $user,
+            'pass' => $pass,
+            'tls' => $tls);
+
+        if ($type === 'jmap') {
+            $imap_list['type'] = 'jmap';
+            $imap_list['hide'] = $hidden;
+            $imap_list['port'] = false;
+            $imap_list['tls'] = false;
+        }
+
+        if (isset($imap_sieve_host) && $imap_sieve_host) {
+            $imap_list['sieve_config_host'] = $imap_sieve_host;
+        }
+
+        Hm_IMAP_List::add($imap_list);
+        $servers = Hm_IMAP_List::dump(false, true);
+        $ids = array_keys($servers);
+        $imap_server_id = array_pop($ids);
+        $server = Hm_IMAP_List::get($imap_server_id, false);
+
+        if ($enableSieve &&
+            isset($imap_sieve_host) &&
+            $context->module_is_supported('sievefilters') &&
+            $context->user_config->get('enable_sieve_filter_setting', true)) {
+            try {
+                list($sieve_host, $sieve_port, $sieve_tls) = parse_sieve_config_host($imap_sieve_host);
+                $client = new \PhpSieveManager\ManageSieve\Client($sieve_host, $sieve_port);
+                $client->connect($user, $pass, $sieve_tls, "", "PLAIN");
```

Comment:
Can we reuse the sieve module here? If needed we can define another method/function to do this. That's too deep sieve integration usage in an imap module which is not recommended as we mix module internals here.

See Hm_Sieve_Client_Factory::init - the 3 lines there almost repeat here - try to reuse that code. You can still catch the exception here and msg accordingly.